src/formatDate.js

```diff
@@ -8,7 +8,76 @@
  * @returns {string}
  */
 function formatDate(date, fromFormat, toFormat) {
-  // write code here
+  const formatDictionary = {};
+  const oldSeparator = fromFormat[3];
+  const newSeparator = toFormat[3];
+  const inputDateParts = date.split(oldSeparator);
+  const newDateParts = [];
+
+  /* Create a dictionary with:
+   * - key - first letter of the part format ('D', 'M' or 'Y')
+   * - value - dictionary with the given 'format' and 'value' of this date part
+   */
+  for (let i = 0; i < 3; i++) {
+    const format = fromFormat[i];
+    const formatCategory = format[0];
+    const value = inputDateParts[i];
+
+    formatDictionary[formatCategory] = {
+      format: format,
+      value: value,
+    };
+  }
+
+  /* For each format part of the 'toFormat', use the first letter as the key
+   * to find the value that should be inserted into the same position and its
+   * initial format.
+   * If the value is the year number, check and if needed update the format.
+   */
+  for (let i = 0; i < 3; i++) {
+    const format = toFormat[i];
+    const formatCategory = format[0];
+    const oldFormat = formatDictionary[formatCategory]['format'];
```

Comment:
You're accessing the 'format' property of the dictionary object without checking if the `formatCategory` key exists. This could lead to potential runtime errors if the `formatCategory` is not one of the expected ones. It's safe in this context due to the controlled input, but adding a check could prevent errors in future changes.